src/BalanceTx.purs

```diff
@@ -0,0 +1,916 @@
+  | CalculateMinFeeFailure' CalculateMinFeeFailure
+  | BuildTxRawFailure' BuildTxRawFailure
+derive instance genericBalanceTxFailure :: Generic BalanceTxFailure _
+
+instance showBalanceTxFailure :: Show BalanceTxFailure where
+  show = genericShow
+
+newtype BalanceTxMFailure = BalanceTxMFailure BalanceTxMFailureReason
+derive instance newtypeBalanceTxMFailure :: Newtype BalanceTxMFailure _
+derive newtype instance showBalanceTxMFailure :: Show BalanceTxMFailure
+
+data BalanceTxMFailureReason
+  = UnknownRequiredSigners
+
+instance showBalanceTxMFailureReason :: Show BalanceTxMFailureReason where
+  show UnknownRequiredSigners = "Unknown required signers."
+
+newtype ReturnAdaChangeFailure
+  = ReturnAdaChangeFailure ReturnAdaChangeFailureReason
+derive instance newtypeReturnAdaChangeFailure
+  :: Newtype ReturnAdaChangeFailure _
+derive newtype instance showReturnAdaChangeFailure
+  :: Show ReturnAdaChangeFailure
+
+data ReturnAdaChangeFailureReason
+  = NotEnoughAdaInputAfterPrebalance Impossible
+  | CouldNotModifyUtxo
+  | CouldNotModifyUtxoHead Impossible
+  | InputAdaDoesNotCoverSingleAdaOutput
+  | ReturnAdaChangeCalculateMinFee CalculateMinFeeFailureReason
+  | ReturnAdaChangeBuildTxRaw BuildTxRawFailureReason
+
+instance showReturnAdaChangeFailureReason
+  :: Show ReturnAdaChangeFailureReason where
+    show (NotEnoughAdaInputAfterPrebalance Impossible) =
+      "Not enough Input Ada to cover output and fees after prebalance - this \
+      \should be **IMPOSSIBLE** if called after loop."
+    show CouldNotModifyUtxo =
+      "Couldn't modify utxo to return change."
+    show (CouldNotModifyUtxoHead Impossible) =
+      "Couldn't modify head utxo to add Ada - this should be **IMPOSSIBLE**."
+    show InputAdaDoesNotCoverSingleAdaOutput =
+      "ReturnAda' does not cover min. utxo requirement for single Ada-only \
+      \output."
+    show (ReturnAdaChangeCalculateMinFee reason) = show reason
+    show (ReturnAdaChangeBuildTxRaw reason) = show reason
+
+newtype AddTxCollateralsFailure
+  = AddTxCollateralsFailure AddTxCollateralsFailureReason
+derive instance newtypeAddTxCollateralsFailure
+  :: Newtype AddTxCollateralsFailure _
+derive newtype instance showAddTxCollateralsFailure
+  :: Show AddTxCollateralsFailure
+
+data AddTxCollateralsFailureReason
+  = CollateralUtxosUnavailable
+
+instance showAddTxCollateralsFailureReason
+  :: Show AddTxCollateralsFailureReason where
+    show CollateralUtxosUnavailable =
+      "There are no utxos to be used as collateral."
+
+newtype ToEitherTransactionInputFailure
+  = ToEitherTransactionInputFailure ToEitherTransactionInputFailureReason
+derive instance newtypeToEitherTransactionInputFailure
+  :: Newtype ToEitherTransactionInputFailure _
+derive newtype instance showToEitherTransactionInputFailure
+  :: Show ToEitherTransactionInputFailure
+
+data ToEitherTransactionInputFailureReason
+  = CannotConvertOutputToTxInput
+
+instance showToEitherTransactionInputFailureReason
```

Comment:
Instance names are optional in PureScript already

Reply:
@ngua something worth deciding for https://github.com/Plutonomicon/cardano-browser-tx/issues/56
I'm not sure what is considered idiomatic but I've always added names, although I should add I've only been learning for a few months.

Reply:
As a personal preference, I would like to omit them. They add a lot of noise in my opinion. I think it's OK for now if there's some discrepancies between different modules when it comes to this, though. We can go back and make everything more uniform later